lib/spack/spack/build_systems/intel.py

```diff
@@ -103,6 +135,711 @@ def _filtered_components(self):
+            else:
+                component = 'compiler'
+
+        d = Prefix(join_path(d, component))
+        debug_print(d)
+        return d
+
+    def relative_path(self, file_or_dir, relative_to=''):
+        if not relative_to:
+            relative_to = self.prefix
+        file_or_dir = os.path.realpath(file_or_dir)
+        relative_to = os.path.realpath(relative_to)
+        p = os.path.relpath(file_or_dir, relative_to)
+        return p
+
+    def component_bin_dir(self, component=None, relative=False):
+        d = self.component_dir(component)
+
+        if sys.platform == 'darwin':
+            # TODO: verify
+            d = d.bin
+        else:
+            if component == 'mpi' or 'intel-mpi' in self.name:
+                d = d.intel64.bin
+            elif component == 'compiler' or 'parallel-studio' in self.name:
+                d = Prefix(ancestor(d)).bin.intel64
+                # NB: psxevars.sh is wrested specially from studio_dir, in
+                # intel-parallel-studio/package.py:file_to_source
+            else:
+                d = d.bin
+
+        if relative:
+            d = self.relative_path(d)
+
+        debug_print(d)
+        return d
+
+    def component_lib_dir(self, component=None, relative=False):
+        # Provide starting directory for find_libraries() and
+        # SPACK_COMPILER_EXTRA_RPATHS.
+        d = self.component_dir(component)
+
+        if sys.platform == 'darwin':
+            d = d.lib
+        else:
+            if component == 'mpi' or 'intel-mpi' in self.name:
+                d = d.intel64.lib
+            else:
+                d = d.lib.intel64
+            # A bit weird, but  I'm sure there are good reasons for it.
+
+        if relative:
+            d = self.relative_path(d)
+
+        debug_print(d)
+        return d
+
+    @property
+    def component_include_dir(self, component=None, relative=False):
+        d = self.component_dir(component)
+
+        if component == 'mpi' or 'intel-mpi' in self.name:
+            d = d.intel64
+        d = d.include
+
+        if relative:
+            d = self.relative_path(d)
+
+        debug_print(d)
+        return d
+
+    @property
+    def omp_libs(self):
+        # Supply LibraryList for linking OpenMP
+
+        # FIXME  if sys.platform == 'darwin' ....
+
+        # TODO?  Across Spack packages for "client" applications, are variants
+        #   named consistently enough to determine if OpenMP is even needed?
+
+        if '%intel' in self.spec:
+            omp_libnames = ['libiomp5']
+
+            # Note about search root: For MKL, the directory
+            # "$MKLROOT/../compiler" will be present even for an MKL-only
+            # product installation (as opposed to one being ghosted via
+            # packages.yaml), specificially to provide the 'iomp5' libs.
+
+            omp_libs = find_libraries(
+                omp_libnames,
+                root=self.component_lib_dir(component='compiler'),
```

Comment:
that's quite difficult to track down what this thing is, previously it was simply
```
prefix.compilers_and_libraries.linux.lib.intel64
```
on linux and 
```
prefix.lib
```
on macOS.

Do you really need the complexity here?

Reply:
### Scope
> Do you really need the complexity here?

Absolutely. Taming the hardcoded `compilers_and_libraries` and related paths and coralling them into a _single code instance_ was my prime reason for undertaking this effort. See the comments in product_dir() for the rationale.

The files in branch develop currently having `compilers_and_libraries` are:

```
$ find . -name \*.py | xargs grep compilers_and_libraries
./var/spack/repos/builtin/packages/intel-mpi/package.py:        mpi_root = self.prefix.compilers_and_libraries.linux.mpi.lib64
./var/spack/repos/builtin/packages/intel-mpi/package.py:        mpi_root = self.prefix.compilers_and_libraries.linux.mpi.include64
./var/spack/repos/builtin/packages/intel-mpi/package.py:        bindir = self.prefix.compilers_and_libraries.linux.mpi.intel64.bin
./var/spack/repos/builtin/packages/intel-mpi/package.py:           $ source compilers_and_libraries/linux/mpi/intel64/bin/mpivars.sh
./var/spack/repos/builtin/packages/intel-mpi/package.py:            self.prefix.compilers_and_libraries.linux.mpi.intel64.bin,
./var/spack/repos/builtin/packages/intel/package.py:            join_path('compilers_and_libraries', 'linux', 'bin',
./var/spack/repos/builtin/packages/intel/package.py:            bin_dir = join_path(self.prefix, 'compilers_and_libraries',
./var/spack/repos/builtin/packages/intel/package.py:            lib_dir = join_path(self.prefix, 'compilers_and_libraries',
./var/spack/repos/builtin/packages/intel-mkl/package.py:                    omp_root = prefix.compilers_and_libraries.linux.lib.intel64
./var/spack/repos/builtin/packages/intel-mkl/package.py:            mkl_root = prefix.compilers_and_libraries.linux.mkl.lib.intel64
./var/spack/repos/builtin/packages/intel-mkl/package.py:            self.prefix.compilers_and_libraries.linux.mkl.lib.intel64
./var/spack/repos/builtin/packages/intel-mkl/package.py:            mkl_lib = self.prefix.compilers_and_libraries.linux.mkl.lib.intel64
./var/spack/repos/builtin/packages/intel-mkl/package.py:            mkl_root = self.prefix.compilers_and_libraries.linux.mkl
./var/spack/repos/builtin/packages/intel-parallel-studio/package.py:                omp_root = prefix.compilers_and_libraries.linux.lib.intel64
./var/spack/repos/builtin/packages/intel-parallel-studio/package.py:        mkl_root = prefix.compilers_and_libraries.linux.mkl.lib.intel64
./var/spack/repos/builtin/packages/intel-parallel-studio/package.py:        mkl_root = self.prefix.compilers_and_libraries.linux.mkl.lib.intel64
./var/spack/repos/builtin/packages/intel-parallel-studio/package.py:        mpi_root = self.prefix.compilers_and_libraries.linux.mpi.lib64
./var/spack/repos/builtin/packages/intel-parallel-studio/package.py:        mpi_root = self.prefix.compilers_and_libraries.linux.mpi.include64
./var/spack/repos/builtin/packages/intel-parallel-studio/package.py:            mkl_root = self.prefix.compilers_and_libraries.linux.mkl.lib.intel64  # noqa
./var/spack/repos/builtin/packages/intel-parallel-studio/package.py:            bindir = self.prefix.compilers_and_libraries.linux.mpi.intel64.bin
```

I addressed `intel-mkl`, `intel-mpi`, and `intel-parallel-studio` within this PR, but not yet `intel`, which is in fact a mere subset of  `intel-parallel-studio` and should be easy to include.

I understand your concern about the PR size. But consider:

* The conceptual change in directory handling necessarily affects all child packages and can easily be leveraged for them.
* Benefit of unified code (see  `def blas_libs()` below).
* Liberal comments (too many?)

_Factoring out MPI_ from the PR as you hint at would not change the affected line counts that much since `intel-mpi` has been the smallest of the affected packages.

### Original location
The `omp_libs` code was previously located within `def blas_libs()` in two different files, where it in fact it had suffered from (presumably undesirable) code drift:

```diff
--- var/spack/repos/builtin/packages/intel-parallel-studio/package.py   2018-03-13 14:57:28.000000000 -0500
+++ var/spack/repos/builtin/packages/intel-mkl/package.py       2018-03-13 14:55:45.000000000 -0500
…
@@ -185,20 +101,26 @@
                 mkl_threading = ['libmkl_intel_thread']
                 omp_threading = ['libiomp5']
 
-                omp_root = prefix.compilers_and_libraries.linux.lib.intel64
+                if sys.platform != 'darwin':
+                    omp_root = prefix.compilers_and_libraries.linux.lib.intel64
+                else:
+                    omp_root = prefix.lib
                 omp_libs = find_libraries(
                     omp_threading, root=omp_root, shared=shared)
             elif '%gcc' in spec:
                 mkl_threading = ['libmkl_gnu_thread']
 
                 gcc = Executable(self.compiler.cc)
-                omp_libs = gcc('--print-file-name', 'libgomp.{0}'.format(
+                libgomp = gcc('--print-file-name', 'libgomp.{0}'.format(
                     dso_suffix), output=str)
-                omp_libs = LibraryList(omp_libs)
+                omp_libs = LibraryList(libgomp)
 
         # TODO: TBB threading: ['libmkl_tbb_thread', 'libtbb', 'libstdc++']
 
-        mkl_root = prefix.compilers_and_libraries.linux.mkl.lib.intel64
+        if sys.platform != 'darwin':
+            mkl_root = prefix.compilers_and_libraries.linux.mkl.lib.intel64
+        else:
+            mkl_root = prefix.mkl.lib
 
         mkl_libs = find_libraries(
             mkl_integer + mkl_threading + ['libmkl_core'],
…
```